ext-src/packages/npm/NpmUtils.ts

```diff
@@ -13,47 +13,166 @@
  * limitations under the License.
  */
-import * as path from "path";
+import * as path from 'path';
 import * as fs from "fs";
 import * as _ from "lodash";
 
 import { exec } from "../../utils/exec";
 import { PackageDependenciesHelper } from "../PackageDependenciesHelper";
 import { NpmPackage } from "./NpmPackage";
+import { NPM_SHRINKWRAP_JSON, YARN_LOCK, PACKAGE_LOCK_JSON } from "./NpmScanType";
+
+const NPM_SHRINKWRAP_COMMAND = `npm shrinkwrap`;
+const YARN_LIST_COMMAND = `yarn list`;
+const NPM_LIST_COMMAND = `npm list`;
 
 export class NpmUtils {
-  public async getDependencyArray(): Promise<Array<NpmPackage>> {
+  public async getDependencyArray(manifestType: string): Promise<Array<NpmPackage>> {
     try {
-      const npmShrinkwrapFilename = path.join(
-        PackageDependenciesHelper.getWorkspaceRoot(),
-        "npm-shrinkwrap.json"
-      );
-      if (!fs.existsSync(npmShrinkwrapFilename)) {
-        let { stdout, stderr } = await exec("npm shrinkwrap", {
+      if (manifestType === YARN_LOCK) {
+        let {stdout, stderr} = await exec(YARN_LIST_COMMAND, {
+          cwd: PackageDependenciesHelper.getWorkspaceRoot()
+        });
+
+        if (stdout != "" && stderr == "") {
+          return Promise.resolve(this.parseYarnList(stdout));
+        }
+      } else if (manifestType === NPM_SHRINKWRAP_JSON) {
+        let { stdout, stderr } = await exec(NPM_SHRINKWRAP_COMMAND, {
           cwd: PackageDependenciesHelper.getWorkspaceRoot()
         });
-        let npmShrinkWrapFile = "npm-shrinkwrap.json";
+        let npmShrinkWrapFile = NPM_SHRINKWRAP_JSON;
         let shrinkWrapSucceeded =
           stdout || stderr.search(npmShrinkWrapFile) > -1;
         if (!shrinkWrapSucceeded) {
-          return Promise.reject("Unable to run npm shrinkwrap");
+          return Promise.reject(`Unable to run ${NPM_SHRINKWRAP_COMMAND}`);
         }
+        let obj = JSON.parse(fs.readFileSync(path.join(PackageDependenciesHelper.getWorkspaceRoot(), NPM_SHRINKWRAP_JSON), "utf8"));
+        
+        return Promise.resolve(this.flattenAndUniqDependencies(obj));
+      } else if (manifestType === PACKAGE_LOCK_JSON) {
+        let {stdout, stderr} = await exec(NPM_LIST_COMMAND, {
+          cwd: PackageDependenciesHelper.getWorkspaceRoot()
+        });
+
+        if (stdout != "" && stderr == "") {
+          return Promise.resolve(this.parseNpmList(stdout));
+        }
+      } else {
+        return Promise.reject(`No valid command supplied, have you implemented it? Manifest type supplied: ${manifestType}`);
       }
-      //read npm-shrinkwrap.json
-      let obj = JSON.parse(fs.readFileSync(npmShrinkwrapFilename, "utf8"));
-      
-      return Promise.resolve(this.flattenAndUniqDependencies(obj));
     } catch (e) {
       return Promise.reject(
-        "npm shrinkwrap failed, try running it manually to see what went wrong:" +
-          e.message
+        `${manifestType} read failed, try running it manually to see what went wrong:` +
```

Comment:
trivial: mix of concat and templating, would this work instead?`${manifestType} read failed, try running it manually to see what went wrong: ${e.stderr}`

Reply:
Very nice catch, @allenhsieh !

Reply:
In: 1c7f218